src/metatrain/experimental/phace/modules/precomputations.py

```diff
@@ -0,0 +1,252 @@
+            (2 * l + 1) for l in range(l_max + 1)  # noqa: E741
+        ]
+        if use_sphericart:
+            self.spherical_harmonics = SphericalHarmonicsSphericart(l_max)
+        else:
+            self.spherical_harmonics = SphericalHarmonicsNoSphericart(l_max)
+
+    def forward(
+        self,
+        positions,
+        cells,
+        species,
+        cell_shifts,
+        pairs,
+        structure_pairs,
+        structure_offsets,
+    ):
+        cartesian_vectors = get_cartesian_vectors(
+            positions,
+            cells,
+            species,
+            cell_shifts,
+            pairs,
+            structure_pairs,
+            structure_offsets,
+        )
+
+        bare_cartesian_vectors = cartesian_vectors.values.squeeze(dim=-1)
+        r = torch.sqrt((bare_cartesian_vectors**2).sum(dim=-1))
+
+        spherical_harmonics = self.spherical_harmonics(
+            bare_cartesian_vectors
+        )  # Get the spherical harmonics
+        spherical_harmonics = spherical_harmonics * (4.0 * torch.pi) ** (
+            0.5
+        )  # normalize them
+        spherical_harmonics = torch.split(
+            spherical_harmonics, self.spherical_harmonics_split_list, dim=1
+        )  # Split them into l chunks
+
+        spherical_harmonics_blocks = [
+            TensorBlock(
+                values=spherical_harmonics_l.unsqueeze(-1),
+                samples=cartesian_vectors.samples,
+                components=[
+                    Labels(
+                        names=("o3_mu",),
+                        values=torch.arange(
+                            start=-l, end=l + 1, dtype=torch.int32
+                        ).reshape(2 * l + 1, 1),
+                    ).to(device=cartesian_vectors.values.device)
+                ],
+                properties=Labels(
+                    names=["_"],
+                    values=torch.zeros(
+                        1, 1, dtype=torch.int32, device=cartesian_vectors.values.device
+                    ),
+                ),
+            )
+            for l, spherical_harmonics_l in enumerate(spherical_harmonics)  # noqa: E741
+        ]
+        spherical_harmonics_map = TensorMap(
+            keys=Labels(
+                names=["o3_lambda"],
+                values=torch.arange(
+                    len(spherical_harmonics_blocks), device=r.device
+                ).reshape(len(spherical_harmonics_blocks), 1),
+            ),
+            blocks=spherical_harmonics_blocks,
+        )
+
+        r_block = TensorBlock(
+            values=r.unsqueeze(-1),
+            samples=cartesian_vectors.samples,
+            components=[],
+            properties=Labels(
+                names=["_"],
+                values=torch.zeros(1, 1, dtype=torch.int32, device=r.device),
+            ),
+        )
+
+        return r_block, spherical_harmonics_map
+
+
+def get_cartesian_vectors(
+    positions, cells, species, cell_shifts, pairs, structure_pairs, structure_offsets
+):
```

Comment:
Why not use the metatensor NL directly? It already has everything required, except the `species_center`/`species_neighbor` metadata

Reply:
Because it's not as fast if there are multiple structures in the batch
The metadata needs to be changed a bit since successive structures need to have their indices changed. It does come from the neighbor lists before being fed into this function

Reply:
ok, could you mention this in a comment? So the next person coming does not try to change it back without adding a bunch of benchmarks